mantidproject/mantid
# Abins: tenth-order phonon scattering approximated by self-convolution

**Description of work.**

This implements a feature requested by instrument scientists and described in the OCLIMAX paper https://doi.org/10.1021/acs.jctc.8b01250 . Rather than compute intensities based on eigenvector information, high-order spectra are obtained by repeatedly convolving the highest-available spectrum with the fundamental spectrum, on a per-atom basis.

It is not made explicit in the OCLIMAX paper (or cited book https://www.its.caltech.edu/~matsci/btfgrp/Inelastic_Neutron_Book.pdf) how the convolution kernel should be scaled. We cannot rely on the native scattering units as these are not dimensionless. (E.g. if you choose a small enough unit that the sum over the kernel exceeds unity, then the spectrum would _increase_ in intensity with each order, which is not physically plausible.) I have therefore chosen to normalise the convolution kernel to 1, such that each order has the same total intensity, and provided an adjustable scaling factor as abins.parameters.autoconvolution['scale'].

The paper also suggests that this step should be applied before the Debye-Waller factor, which is not the case in the current implementation. (The Debye-Waller term is included directly in the intensity calculations; this would have to be decoupled.)

In order to avoid working with large collections of frequency/intensity data, convolution is performed on a discrete grid. However, performing multi-step calculations on a discrete grid is problematic as errors can propagate; this manifests as a slight broadening effect as some events are put into the wrong bins. (e.g. with a bin width of 0.01, the combination of events at 0.999 and 1.002 _should_ end up in the bin (1.00-1.01); instead the events are discretised to 0.995 and 1.005, summed to 2.000 and rounded down to the bin (0.99-1.00). Another event might be placed in the 1.00-1.01 bin correctly; what would have been a large combined intensity has now been broadened over two bins. In the next convolution step this 2-bins-wide peak will go on to broaden other events, where it _should_ only create a 1-bin-wide image.)

This error can be limited by working on a finer grid, before rebinning to the final output. This rebinning can be performed during the instrumental broadening step. Using a 10x finer grid limits the error to a single bin, but adds some overhead and increases the storage requirements for scattering data before the final broadening step.

Initial testing used the `scipy.signal.convolve` function, but this was found to be unacceptably noisy in FFT mode and unacceptably slow in direct mode. Scipy 1.4 introduces a `scipy.signal.oaconvolve` overlap-add implementation of convolution which is much faster. (Scipy claims 500x, and this is consistent with my experience.) However, not all Mantid platforms ship with Scipy 1.4 yet; can we request that update? Otherwise I need to replicate the work here - or duplicate it if licensing allows?

**To test:**


- Unit tests have been added for the new SData features
- [x] including a test for the autoconvolution routine
- Abins systemtest should check that nothing has blown up at a high level. (Note that the data was slightly changed; there are plots below to illustrate that the differences are small.)
- Try running Abins a few times with different settings to ensure that the cache invalidation is handling the new feature ok.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
